File: frontend/src/types/adapter.ts
// Adapter 适配器相关类型定义

// 认证类型
export type AuthType = 'oauth2' | 'password';

// 适配器配置
export interface Adapter {
  id: number;
  name: string;           // 适配器标识：gmail, graph, imap
  display_name: string;   // 显示名称：Gmail API, Microsoft Graph, IMAP
  auth_type: AuthType;    // 认证类型：oauth2, password
  description?: string;   // 描述
  is_enabled: boolean;    // 是否启用
  created_at: string;
  updated_at: string;
}

// Provider-Adapter 关联（多对多）
export interface ProviderAdapter {
  provider_id: number;
  adapter_id: number;
  priority: number;       // 优先级，0 为最高
  adapter?: Adapter;      // 预加载的适配器信息
}

// 适配器响应（API 返回格式）
export interface AdapterResponse {
  id: number;
  name: string;
  display_name: string;
  auth_type: AuthType;
  description?: string;
  is_enabled: boolean;
  created_at: string;
  updated_at: string;
}

// 适配器列表响应
export interface AdapterListResponse {
  success: boolean;
  data: AdapterResponse[];
}

// 适配器 API 响应包装
export interface AdapterApiResponse<T = AdapterResponse> {
  success: boolean;
  data: T;
  message?: string;
  error?: string;
}

// 预定义的适配器名称常量
export const AdapterNames = {
  Gmail: 'gmail',
  Graph: 'graph',
  IMAP: 'imap',
} as const;

export type AdapterName = typeof AdapterNames[keyof typeof AdapterNames];

// 适配器显示名称映射
export const AdapterDisplayNames: Record<string, string> = {
  gmail: 'Gmail API (OAuth2)',
  graph: 'Microsoft Graph (OAuth2)',
  imap: 'IMAP (密码认证)',
};

// 工具函数：获取适配器显示名称
export const getAdapterDisplayName = (name: string): string => {
  return AdapterDisplayNames[name] || name;
};

// 工具函数：判断适配器是否为 OAuth2 类型
export const isOAuth2Adapter = (adapter: Adapter | AdapterResponse): boolean => {
  return adapter.auth_type === 'oauth2';
};

// 工具函数：判断适配器名称是否为 OAuth2 类型
export const isOAuth2AdapterName = (name: string): boolean => {
  return name === AdapterNames.Gmail || name === AdapterNames.Graph;
};

File: backend/internal/model/adapter.go
package model

import (
	"time"
)

// Adapter 邮箱协议适配器模型
// 管理不同的邮箱协议实现（Gmail API、Microsoft Graph、IMAP 等）
type Adapter struct {
	ID          int64     `gorm:"primaryKey" json:"id"`
	Name        string    `gorm:"uniqueIndex;size:50;not null" json:"name"` // 适配器唯一标识 (gmail/graph/imap)
	DisplayName string    `gorm:"size:100;not null" json:"display_name"`    // 显示名称
	AuthType    string    `gorm:"size:20;not null" json:"auth_type"`        // 认证类型 (oauth2/password)
	Description string    `gorm:"type:text" json:"description"`             // 描述信息
	IsEnabled   bool      `gorm:"default:true" json:"is_enabled"`           // 是否启用
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// TableName 指定表名
func (Adapter) TableName() string {
	return "adapters"
}

// AdapterAuthType 认证类型常量
const (
	AdapterAuthTypeOAuth2   = "oauth2"   // OAuth2 授权
	AdapterAuthTypePassword = "password" // 密码/授权码
	AdapterAuthTypeToken    = "token"    // Token 认证（用于 WebAPI 适配器）
)

// AdapterName 适配器名称常量
const (
	AdapterNameGmail  = "gmail"  // Gmail API 适配器
	AdapterNameGraph  = "graph"  // Microsoft Graph 适配器
	AdapterNameIMAP   = "imap"   // 通用 IMAP 适配器
	AdapterNameWebAPI = "webapi" // 通用 Web API 适配器
)

// IsOAuth2 检查是否为 OAuth2 认证类型
func (a *Adapter) IsOAuth2() bool {
	return a.AuthType == AdapterAuthTypeOAuth2
}

// IsPassword 检查是否为密码认证类型
func (a *Adapter) IsPassword() bool {
	return a.AuthType == AdapterAuthTypePassword
}

// IsToken 检查是否为 Token 认证类型
func (a *Adapter) IsToken() bool {
	return a.AuthType == AdapterAuthTypeToken
}

// Validate 验证适配器配置的有效性
func (a *Adapter) Validate() error {
	if a.Name == "" {
		return ErrValidation("name", "适配器名称不能为空")
	}

	if a.DisplayName == "" {
		return ErrValidation("display_name", "显示名称不能为空")
	}

	if a.AuthType == "" {
		return ErrValidation("auth_type", "认证类型不能为空")
	}

	// 验证认证类型
	validAuthTypes := map[string]bool{
		AdapterAuthTypeOAuth2:   true,
		AdapterAuthTypePassword: true,
		AdapterAuthTypeToken:    true,
	}
	if !validAuthTypes[a.AuthType] {
		return ErrValidation("auth_type", "认证类型必须是 oauth2、password 或 token")
	}

	return nil
}

// AdapterResponse 用于 API 响应的 Adapter 结构
type AdapterResponse struct {
	ID          int64     `json:"id"`
	Name        string    `json:"name"`
	DisplayName string    `json:"display_name"`
	AuthType    string    `json:"auth_type"`
	Description string    `json:"description"`
	IsEnabled   bool      `json:"is_enabled"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// ToResponse 将 Adapter 转换为 AdapterResponse
func (a *Adapter) ToResponse() *AdapterResponse {
	return &AdapterResponse{
		ID:          a.ID,
		Name:        a.Name,
		DisplayName: a.DisplayName,
		AuthType:    a.AuthType,
		Description: a.Description,
		IsEnabled:   a.IsEnabled,
		CreatedAt:   a.CreatedAt,
		UpdatedAt:   a.UpdatedAt,
	}
}
